### PR TITLE
add email masking support

### PR DESCRIPTION
**Pull request by azraaroos10**
## Purpose
Implementation for [issue-7711](https://github.com/wso2/product-is/issues/7711)

This PR provides the above described capability as follows.
1. `EmailAddressRegex` should be enabled.
2. Provide the regex to show the email in UI under `EmailAddressRegex`.

Refer below for the sample configurations configured in deployment.toml file (IS 5.9.0).
````
[authentication.authenticator.email_otp.parameters]
EMAILOTPAuthenticationEndpointURL = "https://localhost:9443/emailotpauthenticationendpoint/emailotp.jsp"
EmailOTPAuthenticationEndpointErrorPage = "https://localhost:9443/emailotpauthenticationendpoint/emailotpError.jsp"
EmailAddressRequestPage = "https://localhost:9443/emailotpauthenticationendpoint/emailAddress.jsp"
usecase = "association"
secondaryUserstore = "primary"
EMAILOTPMandatory = false
sendOTPToFederatedEmailAttribute = false
federatedEmailAttributeKey = "email"
EmailOTPEnableByUserClaim = true
CaptureAndUpdateEmailAddress = true
showEmailAddressInUI = true
useEventHandlerBasedEmailSender = true
emailAddressRegex = '(?&lt;=.{1}).(?=.*@)'
````

With the above implementation and the configurations, the email address will be shown in the UI as follows.
**Enter the code sent to your email ID:`a**********@gmail.com`**
Further, if the regex: `(?&lt;=.)[^@](?=[^@]*?@)|(?:(?&lt;=@.)|(?!^)\\G(?=[^@]*$)).(?=.*\\.)` [1] is used, the same email address will be shown as `a**********@g****.com`.

Basically, it replaces the matched word for the regex with asterisks.
````
email = email.replaceAll(emailAddressRegex, "*");
````

[1] https://stackoverflow.com/questions/43003138/regular-expression-for-email-masking

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 1]
```diff
@@ -1706,6 +1710,27 @@ private boolean isShowEmailAddressInUIEnable(AuthenticationContext context, Map<
...
+     * @return emailAddressRegex
+     */
+    private String getEmailAddressRegex(AuthenticationContext context, Map<String, String> parametersMap) {
+
+        String emailAddressRegex = null;
+        String tenantDomain = context.getTenantDomain();
+        Object propertiesFromLocal = context.getProperty(IdentityHelperConstants.GET_PROPERTY_FROM_REGISTRY);
+        if ((propertiesFromLocal != null || tenantDomain.equals(EmailOTPAuthenticatorConstants.SUPER_TENANT)) &&
+                parametersMap.containsKey(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX)) {
+            emailAddressRegex = parametersMap.get(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX);
```
Comment (ruwanta) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:1733]: need to have debug logs

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 2]
```diff
@@ -1706,6 +1710,27 @@ private boolean isShowEmailAddressInUIEnable(AuthenticationContext context, Map<
...
+        } else if ((context.getProperty(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX)) != null) {
+            emailAddressRegex = String.valueOf(context.getProperty(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX));
```
Comment (ruwanta) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:1738]: need to have debug logs

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 3]
```diff
@@ -927,6 +927,10 @@ private void redirectToEmailOTPLoginPage(HttpServletResponse response, Authentic
             }
             String url = getRedirectURL(emailOTPLoginPage, queryParams);
             if (isShowEmailAddressInUIEnable(context, emailOTPParameters)) {
+                String emailAddressRegex = getEmailAddressRegex(context, emailOTPParameters);
+                if (StringUtils.isNotEmpty(emailAddressRegex)) {
+                    email = email.replaceAll(emailAddressRegex, "*");
+                }
```
Comment (ruwanta) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:936]: need to have debug logs for emailAddressRegex null case

**Review by ruwanta — COMMENTED** [review]
Need few debug logs

**Comment by ruwanta** [comment]
Hi @azraaroos10,
Would you able to improve the code by adding few debug logs when on each if condition
 please follow
if(log.isDebugEnabled() {
   log.debug("Some meaningful log")
}

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 2 above)
Comment (azraaroos10) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:1738]: Fixed

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 3 above)
Comment (azraaroos10) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:936]: Fixed

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java (hunk 1 above)
Comment (azraaroos10) [reply component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java:1733]: Fixed

**Comment by azraaroos10** [comment]
Hi @ruwanta , 
Thank you for reviewing the PR. Added the debug logs as requested. Pls, check.

**Review by ruwanta — APPROVED** [review]

**Comment by ruwanta** [comment]
Hi @azraaroos10 ,
Thanks for the changes. Change looks good now.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 4]
```diff
@@ -927,6 +927,16 @@ private void redirectToEmailOTPLoginPage(HttpServletResponse response, Authentic
...
             String url = getRedirectURL(emailOTPLoginPage, queryParams);
             if (isShowEmailAddressInUIEnable(context, emailOTPParameters)) {
+                String emailAddressRegex = getEmailAddressRegex(context, emailOTPParameters);
+                if (StringUtils.isNotEmpty(emailAddressRegex)) {
+                    if (log.isDebugEnabled()){
+                        log.debug("Email address regex defined. Masking the email address using the regex.");
+                    }
+                    email = email.replaceAll(emailAddressRegex, "*");
+                }
+                else if (log.isDebugEnabled()){
```
Comment (senthalan) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]:
```suggestion
                else if (log.isDebugEnabled()) {
```

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 5]
```diff
@@ -1706,6 +1716,33 @@ private boolean isShowEmailAddressInUIEnable(AuthenticationContext context, Map<
...
+     */
+    private String getEmailAddressRegex(AuthenticationContext context, Map<String, String> parametersMap) {
+
+        String emailAddressRegex = null;
+        String tenantDomain = context.getTenantDomain();
+        Object propertiesFromLocal = context.getProperty(IdentityHelperConstants.GET_PROPERTY_FROM_REGISTRY);
+        if ((propertiesFromLocal != null || tenantDomain.equals(EmailOTPAuthenticatorConstants.SUPER_TENANT)) &&
+                parametersMap.containsKey(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX)) {
+            emailAddressRegex = parametersMap.get(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX);
+            if(log.isDebugEnabled()){
```
Comment (senthalan) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]:
```suggestion
            if (log.isDebugEnabled()) {
```

File: component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java [hunk 6]
```diff
@@ -1706,6 +1716,33 @@ private boolean isShowEmailAddressInUIEnable(AuthenticationContext context, Map<
...
+                log.debug("Getting the email address regex from parameters map: " + emailAddressRegex);
+            }
+        } else if ((context.getProperty(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX)) != null) {
+            emailAddressRegex = String.valueOf(context.getProperty(EmailOTPAuthenticatorConstants.EMAIL_ADDRESS_REGEX));
+            if(log.isDebugEnabled()){
```
Comment (senthalan) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java]:
```suggestion
            if (log.isDebugEnabled()) {
```

**Review by senthalan — COMMENTED** [review]
some formating issues

**Comment by azraaroos10** [comment]
Thank you for reviewing and providing the suggestions. All the suggestions are applied.

**Comment by senthalan** [comment]
Thank you for the Contribution. I am merging this PR. It will be great if you can document this configuration in https://is.docs.wso2.com/en/latest/learn/configuring-email-otp/#configure-the-email-otp-provider. You can find the documentation page, https://github.com/wso2/docs-is/blob/master/en/docs/learn/configuring-email-otp.md
